Project Optimization and Git package support

smitdylan2001 opened the pull request:
PR containing performance optimizations, assembly definition support, and git package support for easy package installs. Tested to work in sample scenes.

# Assembly definitions:
Zed core
Zed Editor
Zed Examples (installed with Samples from package manager)

# Git support:
After merging install from https://github.com/stereolabs/zed-unity.git?path=/ZEDCamera/Assets
Current install URL: https://github.com/smitdylan2001/zed-unity.git?path=/ZEDCamera/Assets#ProjectOptimization

# Performance optimizations:
- Merge set .position and .rotation into 1 call (40% lower frametime)
- Merge set .LocalPosition and .localRotation into 1 call with conditional compilation (with assembly definitions)(40% lower frametime)
- Merge get .position and .rotation into 1 call
- Merge bool assignment from if-else to direct assignment
- Merge bool return values from if-else to direct return
- Remove unnecessary ToString() calls
- Reorder vector calculations for better performance
- Optimized string comparisons (reduced garbage allocation + faster comparison)
- Optimized stringBuilder calls
- Removed unneeded .gameObject calls on GameObject values
- Use .AddRange instead of .Add in a loop
- Use string.length instead of comparing to empty string
- Merge double if statement in big loops to reduce branching
- Merge local position and rotation in ZED positional update (runs every frame, so will be helpful)
- Check for bool before checking for XR Device to save performance with mono rig in ZEDManager

# Todo:
[x] Test with OpenCV (I do not have a license)
[ ] Change install instructions in readme to git URL in package manager

SLJLMacarit commented:
The "Common" folder can not be imported through the package manager, breaking the green screen sample. Furthermore, it's only ever used with the green screen sample. Can you confirm the fix would be to just move the folder inside of the green screen sample folder? That's something we can do ourselves afterwards.

SLJLMacarit commented:
Under Assets/Samples/ZED SDK/4.1.0/Green Screen/Shaders/:
- ComputeGreenScreen
- GreenScreen
- GreenScreen_URP
- YUV

These shaders have a broken reference:
Couldn't open include file '../../../SDK/Helpers/Shaders/ZED_Utils.cginc'.

This is probably due to the separation of the package files and the samples files.
Can this be fixed?

SLJLMacarit commented:
The Plane Detection and VR Plane Detection samples were split in two folders whereas they were in the same one before. This leads to duplicated code and assets. Can they be put back into the same folder?

SLJLMacarit commented:
Tested with OpenCV, no issue 👍

SLJLMacarit commented:
The Mixed Reality Calibration (Packages/com.stereolabs.zed/Tools/Mixed Reality Calibration/Scene/MR Calibration.unity) is not available in the samples, so it can't be opened as is. It must be copied from the package into the project. 
Can you make it into a sample ? Since it's the only thing in the tools folder, getting rid of the tools folder altogether is not a problem.

SLJLMacarit commented on ZEDCamera/Assets/package.json:
```diff
@@ -0,0 +1,106 @@
+      "description": "Print out ArUco markers, put them in view of your ZED, and let the battle of the drones begin. Shows how to easily interface the ZED with OpenCV for marker detection using a variety of included scripts.",
+      "path": "Samples~/OpenCV ArUco Detection"
+    },
+    {
+      "displayName": "Plane Detection",
+      "description": "Run the scene and hold down spacebar to see if you’re looking at a valid surface where a bunny could stand. Release the spacebar and a bunny will fall from the sky and land on that surface with proper physics.",
+      "path": "Samples~/Plane Detection"
+    },
+    {
+      "displayName": "VR Plane Detection",
+      "description": "Throw a plushie as far as possible, the distance is measured where it lands using the Plane Detection. Shows how plane detection can fit into a proper VR game.",
+      "path": "Samples~/Plane Detection"
+    },
```
```suggestion
    {
      "displayName": "Plane Detection",
      "description": "Run the scene and hold down spacebar to see if you’re looking at a valid surface where a bunny could stand. Release the spacebar and a bunny will fall from the sky and land on that surface with proper physics. In VR, throw a plushie as far as possible, the distance is measured where it lands using the Plane Detection. Shows how plane detection can fit into a proper VR game.",
      "path": "Samples~/Plane Detection"
    }
```
I suppose this is how we could merge both Plane Detection samples?

SLJLMacarit reviewed (changes requested):
Current state:

- Assembly definition setup (separate base code, samples, and editor code)
  - Please make the MR Calibration a sample (I suppose it involves adding an assembly def ?).
  - Please fix the broken reference in the shaders (or advise a possible solution).
  - Not necessarily explicitly related to assembly definition, but please move the common folder content into the green screen folder.

- Unity specific code optimizations, mainly focused on optimizing hierarchy calculations (best to do if assembly definitions are implemented to set some custom defines) 
  - Nothing to say, very satisfied of the optimizations. 👍 

- Unity git package setup
  - Please add the option to import all the samples at once.
  - Please merge the VR and non-VR Plane detection samples import.

- Documenting changes etc in git PR + project setup time
  - Very comprehensive list of changes. 👍

smitdylan2001 commented on ZEDCamera/Assets/SDK/Stereolabs.Zed.asmdef:
This is because the feature was pushes halfway through the 2021.3 and 2022.2 cycles. Not doing this would break projects between those versions. We can remove it, but then the projects cannot be 2021.1 or 2022.2 projects (which might be acceptable by now)

smitdylan2001 replied:
If you meant why separate tools and the main zed asmdef, this was done to ensure the code base stays separated. The main SDK should not reference tools built upon it.

smitdylan2001 commented:
> The "Common" folder can not be imported through the package manager, breaking the green screen sample. Furthermore, it's only ever used with the green screen sample. Can you confirm the fix would be to just move the folder inside of the green screen sample folder? That's something we can do ourselves afterwards.

Put the folder in the sample folder, I will test next push



> Under Assets/Samples/ZED SDK/4.1.0/Green Screen/Shaders/:
> 
> * ComputeGreenScreen
> * GreenScreen
> * GreenScreen_URP
> * YUV
> 
> These shaders have a broken reference: Couldn't open include file '../../../SDK/Helpers/Shaders/ZED_Utils.cginc'.
> 
> This is probably due to the separation of the package files and the samples files. Can this be fixed?

Ah I see. I can try to hard link them, so making them reference packages. This will break the development project tho.
An alternative is make a copy of the utils include in the green screen shader folder, or potentially (if this does not break references) have the green screen shaders in the SDK/Helpers/Shaders folder as well. Shaders do not add compile time, so that should be fine.



> Is it difficult to add an "Import All Samples" option to the packages manager?

This is not something which is supported by Unity.
I could make a parent folder with all samples, but if people import all samples, and after that an individual sample that will import the sample twice, causing compilation issues. 
I would advice against it and make it a feature request for Unity



> The Mixed Reality Calibration (Packages/com.stereolabs.zed/Tools/Mixed Reality Calibration/Scene/MR Calibration.unity) is not available in the samples, so it can't be opened as is. It must be copied from the package into the project. Can you make it into a sample ? Since it's the only thing in the tools folder, getting rid of the tools folder altogether is not a problem.

I will add it as a sample and put it under the Sample asmdef in the next push

smitdylan2001 commented:
It would be nice to add the MR calibration in the Samples in the git readme as well. Now it only is in the documentation

smitdylan2001 commented:
Fixed the sample issues and added the mixed reality sample
Also fixed an issue when importing the green screen sample without asmdef support added because the editor reference was lost

Bvallon-sl commented on ZEDCamera/Assets/SDK/Helpers/Scripts/ZEDManager.cs:
```diff
@@ -2044,6 +2044,11 @@ private void ClearRendering()
     /// </summary>
     void Awake()
     {
+        if (UpgradePluginToSRP.UpgradeGreenScreenToSRP(gameObject))
```
`UpgradePluginToSRP.UpgradeCameraToSRP` should be called instead.
